Adding initial version of `goto-inspect`.

NlightNFotis opened the pull request:
This allows to inspect a goto-binary (this means showing the goto-functions, for now at least).

The reason this binary is helpful is that different tools (CBMC, goto-instrument, etc) perform instrumentation before showing the goto-functions, and there's a need to be able to inspect certain properties of a goto-binary without doing instrumentation.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

NlightNFotis committed:
Adding initial version of `goto-inspect`.

This allows to inspect a goto-binary (this means showing the goto-functions,
for now at least).

The reason this binary is helpful is that different tools (CBMC, goto-instrument, etc)
perform instrumentation before showing the goto-functions, and there's a need
to be able to inspect certain properties of a goto-binary without doing instrumentation.

martin-cs commented:
How different is this to just using `goto-instrument --show-goto-programs` ?

martin-cs commented:
Apologies, I hadn't seen https://github.com/diffblue/cbmc/issues/7674, will discuss there.

tautschnig commented:
Please take a look at https://github.com/diffblue/cbmc/pull/7415, which is the most recent new tool that was added and is much more complete in what it takes to add a new tool.

NlightNFotis commented:
Hi Michael (@tautschnig), latest commit should have everything that has been asked.

Let me know if this is closer to what you envisioned.

thomasspriggs commented on src/goto-inspect/CMakeLists.txt:
```diff
@@ -0,0 +1,22 @@
+install(TARGETS goto-inspect DESTINATION ${CMAKE_INSTALL_BINDIR})
+
+# # Man page
+# if(NOT WIN32)
```
Why is this commented out?

NlightNFotis replied:
Good catch - it shouldn't be commented out.

It was in the first commit, when we didn't have a man page at first, but now we do.

thomasspriggs commented on CMakeLists.txt:
```diff
@@ -230,7 +231,9 @@ cprover_default_properties(
     goto-diff-lib
+    goto-inspect
     goto-harness
+    goto-inspect
```
Nitpick: duplication

NlightNFotis replied:
Good catch.

thomasspriggs commented on src/goto-inspect/goto_inspect_parse_options.cpp:
```diff
@@ -0,0 +1,99 @@
+  // This just sets up the defaults (and would interpret options such as --32).
+  config.set(cmdline);
+
+  // Normally we would register language front-ends here but as goto-harness
```
Should "goto-harness" here be "goto-inspect"?

NlightNFotis replied:
Yes, this was a copy-paste fail.

I used the code under `src/goto-harness` as the base for a lot of what you're seeing here, but it seems like I forgot to update some of the references.

Will have another look myself, just to make sure.

thomasspriggs reviewed (commented):
Looks reasonable to me, aside from a few minor issues.

TGWDB commented on doc/man/goto-inspect.1:
```diff
@@ -0,0 +1,33 @@
+.TH GOTO-INSPECT "2" "May 2023" "goto-inspect-5.81.0" "User Commands"
```
Probably not worth pinning a date/version here.

NlightNFotis replied:
[This reference](https://man7.org/linux/man-pages/man7/man-pages.7.html) leads me to believe that the date is necessary in the title line.

It does look like the date we have though doesn't align with the proposed format in the reference I linked. I'm not sure how the tools behave if I leave it with an empty date, but it seems to be diverging from the expected format too much IMO.

TGWDB commented on doc/man/goto-inspect.1:
```diff
@@ -0,0 +1,33 @@
+.BR goto-cc (1)
+.BR goto-instrument (1)
+.SH COPYRIGHT
+2023, Fotis Koutoulakis
```
Copyright is Diffblue I believe.

NlightNFotis replied:
You are right - I copied and hastily changed this from a previous one, so I've made a number of mistakes.

I'm fixing now.

TGWDB commented on src/goto-inspect/goto_inspect_parse_options.cpp:
```diff
@@ -0,0 +1,97 @@
+    return CPROVER_EXIT_SUCCESS;
+  }
+
+  // Before we do anything else, positional arguments needs to be > 1.
+  // (i.e. a filename has been given).
+  if(cmdline.args.size() < 1)
+  {
+    help();
+    throw invalid_command_line_argument_exceptiont{
+      "failed to supply a goto-binary name",
+      "<input goto-binary> <inspection-option>"};
+  }
```
This looks like it should be `> 2` and `< 2` since we require a binary name and another option?

Maybe we also need a code path before this to response to the `--help` option.

NlightNFotis replied:
> This looks like it should be > 2 and < 2 since we require a binary name and another option?

Yes, changed to `> 2`. With the current implementation, if just a binary is given, the behaviour is that the tool doesn't appear to do anything. It's reasonable in my opinion, but not very helpful, as it doesn't indicate to the user whether it succeeded or failed, or what.

> Maybe we also need a code path before this to response to the --help option.

We don't need to do that, as it's handled already by the base class `parse_options_baset::main()`, where it's checking for `h` or `help` before calling `doit()`.

TGWDB commented on src/goto-inspect/goto_inspect_parse_options.cpp:
```diff
@@ -0,0 +1,97 @@
+    return CPROVER_EXIT_SUCCESS;
+  }
+
+  // No options was passed in - erroneously?
+  return CPROVER_EXIT_INCORRECT_TASK;
```
The failure for lack of an option should be caught/reported earlier as in comment above, so this code path should perhaps be unreachable?

NlightNFotis replied:
It should be unreachable - I have added a lengthier comment addressing why the value is there, in order not to confuse people. Let me know if it now makes more sense.

TGWDB commented on src/goto-inspect/goto_inspect_parse_options.cpp:
```diff
@@ -63,7 +63,14 @@ int goto_inspect_parse_optionst::doit()
 
-  // No options was passed in - erroneously?
+  // If an option + binary was provided, the program will have exited gracefully
+  // through different a different branch. If we hit the code below, it means
```
"different a different" -> "a different"

NlightNFotis replied:
Fixed.

kroening commented:
May I suggest that you consider the user interface. In particular, consider swapping the order of the command line arguments.

All command line tools that I can think of have the order "action" and then "file name". E.g., think `git add some_file.cpp`, or maybe `tar tfz tarball.tgz`.

NlightNFotis commented:
Hi @kroening, I'm working on adding tests (didn't want to commit to writing tests before there's conclusion that the approach is not objectionable to people) and during the authoring of tests I have discovered some issues with the interface that led me to rethink the approach.

Will update the PR as soon as my current change is stable and the interface is smoother.

NlightNFotis commented:
Hi @kroening, I've had some more time to play around with this, and it doesn't have any particular dependency on the order of arguments, i.e. both: `goto-inspect --show-goto-functions example.goto` and `goto-inspect example.goto --show-goto-functions` will work.

peterschrammel commented on src/goto-inspect/goto_inspect_parse_options.cpp:
```diff
@@ -0,0 +1,99 @@
+
+  if(
+    cmdline.isset("show-goto-functions") ||
+    cmdline.isset("list-goto-functions"))
```
This alias is not configured in GOTO_INSPECT_OPTIONS.

NlightNFotis replied:
Removed.

peterschrammel reviewed (approved):
The individual commits are not very clean. I suggest to squash them as they all fix up earlier commits of the same PR.

NlightNFotis commented:
Squashed the commits down to one.

kroening commented:
> Hi @kroening, I've had some more time to play around with this, and it doesn't have any particular dependency on the order of arguments, i.e. both: `goto-inspect --show-goto-functions example.goto` and `goto-inspect example.goto --show-goto-functions` will work.

I got thrown off by your comment "Yes, changed to > 2" -- this is outdated?

NlightNFotis commented:
Hi @kroening,

> I got thrown off by your comment "Yes, changed to > 2" -- this is outdated?

Yes. Thomas had asked if I can tighten the error checking there, which I initially did by checking that the `cmdline.args > 2` (1 for the filename and 1 for the option). After some experimenting during the addition of tests, I found out that the initial approach was (more) correct because `cmdline.args` contains the positional arguments of the program, with the optional ones being added to the `cmdline.options` vector (so if you issue, for example, `$ goto-inspect --show-goto-functions example.gb`, the option `--show-goto-functions` won't be present in the `cmdline.args`, and `cmdline.args == 1` holds)

At that point, I changed it to check that we always get one positional argument (based on the fact that the tool, for now, operates on one goto-binary at a time) before we do anything, and dispatches to various inspection options based on them being set and the above precondition (at least 1 positional argument, i.e. 1 goto-binary filename) holds.

If the positional argument doesn't correspond to a goto-binary, we get an error that the program failed to read a goto-binary.
